cmd/kubectl-gadget/collector.go

```diff
@@ -0,0 +1,389 @@
+		traceConfig.UserAgent = restclient.DefaultKubernetesUserAgent()
+
+		traceRestClient, err := restclient.UnversionedRESTClientFor(&traceConfig)
+
+		for _, node := range nodes.Items {
+			if collectorParamNode != "" && node.Name != collectorParamNode {
+				continue
+			}
+
+			trace := &gadgetv1alpha1.Trace{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: subCommand + "-",
+					Namespace:    "gadget",
+					Annotations: map[string]string{
+						GADGET_OPERATION: "start",
+					},
+					Labels: map[string]string{
+						"trace-template-hash": traceID,
+					},
+				},
+				Spec: gadgetv1alpha1.TraceSpec{
+					Node:   node.Name,
+					Gadget: subCommand,
+					Filter: &gadgetv1alpha1.ContainerFilter{
+						Namespace:     namespace,
+						Podname:       collectorParamPodname,
+						ContainerName: collectorParamContainername,
+						Labels:        labelsSelector,
+					},
+					RunMode:    "Manual",
+					OutputMode: "Status",
+				},
+			}
+			createdTrace := &gadgetv1alpha1.Trace{}
+
+			err = traceRestClient.
+				Post().
+				Namespace(trace.ObjectMeta.Namespace).
+				Resource("traces").
+				Body(trace).
+				Do(context.TODO()).
+				Into(createdTrace)
+			if err != nil {
+				contextLogger.Fatalf("Error creating trace on node %s: %q", node.Name, err)
+			}
+		}
+
+		var listTracesOptions = metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("trace-template-hash=%s", traceID),
+			FieldSelector: fields.Everything().String(),
+		}
+
+		// Wait until results are ready and fetch all results
+		// TODO: use a watcher to avoid looping client-side
+		//
+		// watch, err := traceRestClient.
+		//	Get().
+		//	Namespace("gadget").
+		//	Resource("traces").
+		//	VersionedParams(&listTracesOptions, scheme.ParameterCodec).
+		//	Watch(context.TODO())
+		// if err != nil {
+		//	contextLogger.Fatalf("Error waiting for traces: %q", err)
+		// }
+		// for event := range watch.ResultChan() {
+		//	fmt.Printf("Event: %v\n", event)
+		//	if data, ok := event.Object.(*metav1.Status); ok {
+		//		contextLogger.Infof("watcher status: %s", data.Message)
+		//	} else if t, ok := event.Object.(*gadgetv1alpha1.Trace); ok {
+		//		fmt.Printf("Got: %v\n", t)
+		//	} else {
+		//		contextLogger.Fatalf("Error waiting for traces: got unexpected %v", event)
+		//	}
+		// }
+
+		var results gadgetv1alpha1.TraceList
+		start := time.Now()
+		for {
+			results = gadgetv1alpha1.TraceList{}
+			err = traceRestClient.
+				Get().
+				Namespace("gadget").
+				Resource("traces").
+				VersionedParams(&listTracesOptions, scheme.ParameterCodec).
+				Do(context.TODO()).
+				Into(&results)
+			if err != nil {
+				contextLogger.Fatalf("Error getting traces: %q", err)
+			}
+			again := false
+			for _, i := range results.Items {
+				if i.Status.State != "Completed" {
+					again = true
+				}
+			}
+			if again {
+				elapsed := time.Now().Sub(start)
+				if elapsed > 2*time.Second {
+					contextLogger.Fatalf("Error waiting for traces: timeout: %v", elapsed)
```

Comment:
I think this leaks the tracers, the delete logic should still be called.

Reply:
I'm also wondering if this should just exit here or if it should print the data it was able to collect and a warning indicating that some data is not available because an error occurred. 

Btw, the outputError could be useful here to indicate what went wrong.

Reply:
Fixed.